### PR TITLE
Full traversal in dask.order

### PR DESCRIPTION
**Pull request by mrocklin**
- [x] Tests added / passed
- [x] Passes `flake8 dask`
- [x] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

### CODE REVIEW

**Comment by mrocklin** [comment]
So this test seems to conflict with the following situation:

```python
    def test_avoid_upwards_branching(abcde):
        """
             a1
             |
             a2
             |
             a3    d1
            /  \  /
          b1    c1
          |     |
          b2    c2
                |
                c3
    
        Prefer b1 over c1 because it won't stick around waiting for d1 to complete
        """
        a, b, c, d, e = abcde
        dsk = {(a, 1): (f, (a, 2)),
               (a, 2): (f, (a, 3)),
               (a, 3): (f, (b, 1), (c, 1)),
               (b, 1): (f, (b, 2)),
               (c, 1): (f, (c, 2)),
               (c, 2): (f, (c, 3)),
               (d, 1): (f, (c, 1))}
    
        o = order(dsk)    
```

In this test we want to select `b`, but doing so requires that we start descending from `a`.  The test in this PR asks that we start from short top-nodes because they can be finished more quickly, allowing their children to be released once a long-running co-parent chain starts running.  The test in this PR asks us to start descending from short top-nodes.  The test in this comment asks that we start from long top-nodes.

This brings up the question of not doing top-down depth first, but something that wanders around a bit more or swaps values during traversal.

**Comment by mrocklin** [comment]
OK, there is a solution here that solves a number of problems modestly well.  It is sacrifices perfection in a few cases for "good enough" in more.  The traversal is more complex now and may be more expensive.  I'll have to do some benchmarking.  Here are the visualizations for a couple of troublesome graphs.

dask
gh3055

```python
import matplotlib.pyplot as plt

import dask.array as da
n = 10
x  = da.random.normal(size=(n, 100), chunks=(1, 100))
y  = da.random.normal(size=(n,), chunks=(1,))
xy = (x * y[:, None]).cumsum(axis=0)
xx = (x[:, None, :] * x[:, :, None]).cumsum(axis=0)
beta = da.stack([da.linalg.solve(xx[i], xy[i]) for i in range(xx.shape[0])],
                axis=0)
ey = (x * beta).sum(axis=1)
ey.visualize('dask.png', color='order', node_attr=dict(penwidth='8'),
             optimize_graph=True, cmap=plt.cm.RdBu)


A, B = 10, 99
x = da.random.normal(size=(A, B), chunks=(1, None))
for _ in range(2):
    y = (x[:, None, :] * x[:, :, None]).cumsum(axis=0)
    x = x.cumsum(axis=0)
w = (y * x[:, None]).sum(axis=(1,2))

w.visualize('gh3055.png', color='order', node_attr=dict(penwidth='8'),
            cmap=plt.cm.RdBu)
```

**Comment by mrocklin** [comment]
@eriknw and @jcrist might both find this PR of interest

**Comment by mrocklin** [comment]
Anecdotal timing information

```python
%%time
n = 100
x  = da.random.normal(size=(n, 100), chunks=(1, 100))
y  = da.random.normal(size=(n,), chunks=(1,))
xy = (x * y[:, None]).cumsum(axis=0)
xx = (x[:, None, :] * x[:, :, None]).cumsum(axis=0)
beta = da.stack([da.linalg.solve(xx[i], xy[i]) for i in range(xx.shape[0])],
                axis=0)
ey = (x * beta).sum(axis=1)
CPU times: user 239 ms, sys: 4.21 ms, total: 243 ms
Wall time: 233 ms
```

```python
%time dsk = dict(ey.__dask_graph__())
CPU times: user 74.1 ms, sys: 0 ns, total: 74.1 ms
Wall time: 73.2 ms
```

### Master: 25ms

```python
%timeit _ = order(dsk)
10 loops, best of 3: 25.9 ms per loop
```

### This branch: 39ms

```python
%timeit _ = order(dsk)
10 loops, best of 3: 39.3 ms per loop
```

This is non-trivial, but possibly still worth it.

**Comment by mrocklin** [comment]
Merging this soon if there are no further comments.  Alternatively @jcrist, you're probably the most likely to review this.  Have you had a chance to look it over?

**Comment by jcrist** [comment]
Looking now.

File: dask/tests/test_order.py [hunk 1]
```diff
@@ -213,6 +207,7 @@ def test_stacklimit(abcde):
     ndependencies(dependencies, dependents)
 
 
+@pytest.mark.xfail(reason="Can't please 'em all")
```
Comment (jcrist) [inline dask/tests/test_order.py]: 😆

File: dask/tests/test_order.py [hunk 2]
```diff
@@ -244,10 +239,115 @@ def test_gh_3055():
     dsk = dict(w.__dask_graph__())
     o = order(dsk)
     L = [o[k] for k in w.__dask_keys__()]
-    assert sorted(L) == L[::-1]
+    assert sorted(L) == L[::-1] or sorted(L) == L
```
Comment (jcrist) [inline dask/tests/test_order.py]: Is this non-determinism necessary?

Comment (mrocklin) [reply dask/tests/test_order.py]: Removed

**Comment by mrocklin** [comment]
@jcrist ok to merge on passed tests?

**Comment by jcrist** [comment]
Fine by me.

**Pull request merged**